File: pypto/src/bindings/operation.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file operation.cpp
 * \brief
 */

#include "pybind_common.h"

#include <vector>

using namespace npu::tile_fwk;

namespace pypto {
void bind_operation(py::module &m) {
    m.def(
        "add", [](const Tensor &left, const Tensor &right) { return npu::tile_fwk::Add(left, right); }, "Tensor add.");
    m.def("sub", [](const Tensor left, const Tensor &right) { return npu::tile_fwk::Sub(left, right); }, "Tensor sub.");
    m.def("mul", [](const Tensor left, const Tensor &right) { return npu::tile_fwk::Mul(left, right); }, "Tensor mul.");
    m.def("div", [](const Tensor left, const Tensor &right) { return npu::tile_fwk::Div(left, right); }, "Tensor div.");
    m.def(
        "view",
        [](const Tensor &operand, const std::vector<int64_t> &shapes, const std::vector<int64_t> &offsets) {
            return npu::tile_fwk::View(operand, shapes, offsets);
        },
        "Tensor view.");
    m.def("exp", [](const Tensor &operand) { return npu::tile_fwk::Exp(operand); }, "Tensor exp.");

    m.def(
        "transpose",
        [](const Tensor &operand, const std::vector<int> &transposeShape) {
            return npu::tile_fwk::Transpose(operand, transposeShape);
        },
        "Tensor transpose.");
    m.def("abs", [](const Tensor &operand) { return npu::tile_fwk::Abs(operand); }, "Tensor abs.");
    m.def("reciprocal", [](const Tensor &operand) { return npu::tile_fwk::Reciprocal(operand); }, "Tensor reciprocal.");
    m.def("sqrt", [](const Tensor &operand) { return npu::tile_fwk::Sqrt(operand); }, "Tensor sqrt.");

    m.def(
        "cast",
        [](const Tensor &operand, DataType new_data_type, CastMode mode) {
            return npu::tile_fwk::Cast(operand, new_data_type, mode);
        },
        py::arg("operand"), py::arg("new_data_type"), py::arg("mode") = CAST_NONE, "Tensor cast.");

    m.def(
        "add_s", [](const Tensor &left, const Element &right) { return npu::tile_fwk::AddS(left, right); },
        "Tensor add scalar.");
    m.def(
        "sub_s", [](const Tensor &left, const Element &right) { return npu::tile_fwk::SubS(left, right); },
        "Tensor sub scalar.");
    m.def(
        "mul_s", [](const Tensor &left, const Element &right) { return npu::tile_fwk::MulS(left, right); },
        "Tensor mul scalar.");
    m.def(
        "div_s", [](const Tensor &left, const Element &right) { return npu::tile_fwk::DivS(left, right); },
        "Tensor div scalar.");

    m.def(
        "row_max_single", [](const Tensor &operand, int axis) { return npu::tile_fwk::RowMaxSingle(operand, axis); },
        py::arg("operand"), py::arg("axis") = -1, "Tensor row max single.");

    m.def(
        "row_sum_single", [](const Tensor &operand, int axis) { return npu::tile_fwk::RowSumSingle(operand, axis); },
        py::arg("operand"), py::arg("axis") = -1, "Tensor row sum single.");
    m.def(
        "row_min_single", [](const Tensor &operand, int axis) { return npu::tile_fwk::RowMinSingle(operand, axis); },
        py::arg("operand"), py::arg("axis") = -1, "Tensor row min single.");
    m.def(
        "row_sum_expand", [](const Tensor &operand) { return npu::tile_fwk::RowSumExpand(operand); },
        "Tensor row sum expand.");
    m.def(
        "row_max_expand", [](const Tensor &operand) { return npu::tile_fwk::RowSumSingle(operand); },
        "Tensor row sum expand.");
    m.def("compact", [](const Tensor &operand) { return npu::tile_fwk::Compact(operand); }, "Tensor compact.");
    m.def("indexput", [](const Tensor &src, std::vector<Tensor> indices, const Tensor &values)
        { return npu::tile_fwk::IndexPut(src, indices, values); }, "Tensor indexput.");
    m.def("scatter_element", [](const Tensor &src, const Tensor &idx, const Element &scalar, int axis)
        { return npu::tile_fwk::ScatterElement(src, idx, scalar, axis); }, "Tensor scatter element.");
    m.def("gather_element", [](const Tensor &params, const Tensor &indices, int axis)
        { return npu::tile_fwk::GatherElement(params, indices, axis); }, "Tensor gather element.");
    m.def("gather", [](const Tensor &params, const Tensor &indices, int axis)
        { return npu::tile_fwk::Gather(params, indices, axis); }, "Tensor gather.");
    m.def("duplicate", [](const Tensor &operand) { return npu::tile_fwk::Duplicate(operand); }, "Tensor duplicate.");
    m.def("vector_duplicate", [](const Element &src, DataType dType, std::vector<int64_t> dstShape, 
        std::vector<SymbolicScalar> validShape)
        { return npu::tile_fwk::VectorDuplicate(src, dType, dstShape, validShape); },
        py::arg("src"), py::arg("dType"), py::arg("dstShape"), py::arg("validShape") = std::vector<SymbolicScalar>{},
        "Tensor vector duplicate.");
    m.def("vector_duplicate", [](const SymbolicScalar &src, DataType dType, std::vector<int64_t> dstShape, 
        std::vector<SymbolicScalar> validShape)
        { return npu::tile_fwk::VectorDuplicate(src, dType, dstShape, validShape); },
        py::arg("src"), py::arg("dType"), py::arg("dstShape"), py::arg("validShape") = std::vector<SymbolicScalar>{},
        "Tensor vector duplicate.");
    m.def("reshape", [](const Tensor &input, const std::vector<int64_t> &dstShape,
        const std::vector<SymbolicScalar> validShape) { return npu::tile_fwk::Reshape(input, dstShape, validShape); },
        py::arg("input"), py::arg("dstShape"), py::arg("validShape") = std::vector<SymbolicScalar>{},
        "Tensor reshape.");

    m.def(
        "maximum", [](const Tensor &left, const Tensor &right) { return npu::tile_fwk::Maximum(left, right); },
        py::arg("left"), py::arg("right"), "Tensor maximum.");
    m.def(
        "rms_norm", [](const Tensor &operand) { return npu::tile_fwk::RmsNorm(operand); }, py::arg("operand"),
        "Tensor rms norm.");

    m.def(
        "matmul",
        [](DataType out_type, const Tensor &a, const Tensor &b, bool a_trans, bool b_trans) {
            if (a_trans) {
                if (b_trans) {
                    return npu::tile_fwk::Matrix::Matmul<true, true>(out_type, a, b);
                } else {
                    return npu::tile_fwk::Matrix::Matmul<true, false>(out_type, a, b);
                }
            } else {
                if (b_trans) {
                    return npu::tile_fwk::Matrix::Matmul<false, true>(out_type, a, b);
                } else {
                    return npu::tile_fwk::Matrix::Matmul<false, false>(out_type, a, b);
                }
            }
        },
        py::arg("out_type"), py::arg("a"), py::arg("b"), py::arg("a_trans") = false, py::arg("b_trans") = false,
        "Matrix multiply.");

    m.def(
        "assemble",
        [](const std::vector<std::pair<Tensor, std::vector<int64_t>>> &tensor_int_pairs) {
            return npu::tile_fwk::Assemble(tensor_int_pairs);
        },
        "Tensor::Assemble");
}
} // namespace pypto

File: python/src/tensor_op/pto/operation.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ======================================================================================================================
# pyright: reportReturnType=false
"""
"""
import typing
from typing import Optional, Union

from pto import pto_impl

from .element import Element
from .pto_utils import to_syms
from .symbolic_scalar import SymbolicScalar
from .tensor import Tensor


def _to_base(arg):
    if isinstance(arg, (Tensor, Element, SymbolicScalar)):
        return arg.base()
    elif isinstance(arg, (list, tuple)):
        return [_to_base(a) for a in arg]
    else:
        return arg


def op_wrapper(func):
    def wrapper(*args, **kwargs):
        args = _to_base(args)
        assert isinstance(args, (list, tuple))
        out = func(*args, **kwargs)
        if out is None:
            return None
        elif isinstance(out, pto_impl.Tensor):
            return Tensor.from_base(out)
        else:
            return out
    return wrapper


@op_wrapper
def add(left, right) -> Tensor:
    return pto_impl.add(left, right)


@op_wrapper
def sub(left, right) -> Tensor:
    return pto_impl.sub(left, right)


@op_wrapper
def mul(left, right) -> Tensor:
    return pto_impl.mul(left, right)


@op_wrapper
def div(left, right) -> Tensor:
    return pto_impl.div(left, right)


def view(a, shapes, *args) -> Tensor:
    shapes = [int(s) for s in shapes]
    if len(args) == 1:
        out = pto_impl.view(a.base(), shapes, to_syms(args[0]))
    elif len(args) == 2:
        out = pto_impl.view(a.base(), shapes, to_syms(
            args[0]), to_syms(args[1]))
    else:
        raise RuntimeError("Invalid arguments")
    return Tensor.from_base(out)


def assemble(*args) -> Optional[Tensor]:
    if len(args) == 2:
        tenors = typing.cast(list[pto_impl.Tensor], _to_base(args[0]))
        return Tensor.from_base(pto_impl.assemble(tenors, args[1]))
    elif len(args) == 3:
        pto_impl.assemble(args[0].base(), to_syms(args[1]), args[2].base())
        return None
    else:
        raise RuntimeError("Invalid arguments")


def min(a: 'SymbolicScalar | int', b: 'SymbolicScalar | int') -> 'SymbolicScalar':
    if isinstance(a, int):
        a = SymbolicScalar(a)
    return a.min(b)


def max(a: 'SymbolicScalar | int', b: 'SymbolicScalar | int') -> 'SymbolicScalar':
    if isinstance(a, int):
        a = SymbolicScalar(a)
    return a.max(b)


@op_wrapper
def exp(a) -> Tensor:
    return pto_impl.exp(a)


@op_wrapper
def transpose(a, axis) -> Tensor:
    return pto_impl.transpose(a, axis)


@op_wrapper
def abs(a) -> Tensor:
    return pto_impl.abs(a)


@op_wrapper
def reciprocal(a) -> Tensor:
    return pto_impl.reciprocal(a)


@op_wrapper
def rsqrt(a) -> Tensor:
    return pto_impl.rsqrt(a)


@op_wrapper
def sqrt(a) -> Tensor:
    return pto_impl.sqrt(a)


@op_wrapper
def neg(a) -> Tensor:
    return pto_impl.neg(a)


@op_wrapper
def log(a, base) -> Tensor:
    return pto_impl.log(a, base)


@op_wrapper
def cast(a, dtype, mode) -> Tensor:
    return pto_impl.cast(a, dtype, mode)


@op_wrapper
def add_s(a, b) -> Tensor:
    return pto_impl.add_s(a, b)


@op_wrapper
def sub_s(a, b) -> Tensor:
    return pto_impl.sub_s(a, b)


@op_wrapper
def mul_s(a, b) -> Tensor:
    return pto_impl.mul_s(a, b)


@op_wrapper
def div_s(a, b) -> Tensor:
    return pto_impl.div_s(a, b)


@op_wrapper
def row_max_single(a, axis=-1) -> Tensor:
    return pto_impl.row_max_single(a, axis)


@op_wrapper
def row_sum_single(a, axis=-1) -> Tensor:
    return pto_impl.row_sum_single(a, axis)


@op_wrapper
def matmul(dtype, a, b, **kwargs) -> Tensor:
    a_trans = kwargs.get("a_trans", False)
    b_trans = kwargs.get("b_trans", False)
    c_matrix_nz = kwargs.get("c_matrix_nz", False)
    return pto_impl.matmul(dtype, a, b, a_trans, b_trans, c_matrix_nz)


@op_wrapper
def batch_matmul(dtype, a, b, **kwargs) -> Tensor:
    a_trans = kwargs.get("a_trans", False)
    b_trans = kwargs.get("b_trans", False)
    c_matrix_nz = kwargs.get("c_matrix_nz", False)
    return pto_impl.batch_matmul(dtype, a, b, a_trans, b_trans, c_matrix_nz)
